File: scala/src/main/scala/problems/Problem095MaxRectangleHistogram.scala

package problems
import scala.collection.mutable

object Problem095MaxRectangleHistogram {
  def largestRectangleArea(h:Array[Int]): Int = {
    val st = new mutable.ArrayDeque[Int]()
    var max = 0; var i=0
    while (i<=h.length) {
      val cur = if (i==h.length) 0 else h(i)
      while (st.nonEmpty && cur < h(st.last)) {
        val height = h(st.removeLast())
        val left = if (st.isEmpty) -1 else st.last
        val width = i - left - 1
        max = math.max(max, height * width)
      }
      st.append(i); i+=1
    }
    max
  }
}

File: java/src/main/java/problems/Problem095MaxRectangleHistogram.java
package problems;


import java.util.*;
public class Problem095MaxRectangleHistogram {
    public static int largestRectangleArea(int[] heights){
        int n=heights.length, max=0;
        java.util.Deque<Integer> st=new java.util.ArrayDeque<>();
        for(int i=0;i<=n;i++){
            int h = (i==n)? 0: heights[i];
            while(!st.isEmpty() && h < heights[st.peek()]){
                int height=heights[st.pop()];
                int left = st.isEmpty()? -1: st.peek();
                int width = i - left - 1;
                max = Math.max(max, height*width);
            }
            st.push(i);
        }
        return max;
    }
}

